bson/src/main/org/bson/codecs/configuration/CodecProvider.java

```diff
@@ -34,10 +38,34 @@ public interface CodecProvider {
     /**
      * Get a {@code Codec} using the given context, which includes, most importantly, the Class for which a {@code Codec} is required.
      *
+     * <p>This method is called only if {@link #get(Class, List, CodecRegistry)} is not properly overridden.</p>
+     *
      * @param clazz the Class for which to get a Codec
      * @param registry the registry to use for resolving dependent Codec instances
      * @param <T> the type of the class for which a Codec is required
      * @return the Codec instance, which may be null, if this source is unable to provide one for the requested Class
      */
     <T> Codec<T> get(Class<T> clazz, CodecRegistry registry);
+
+    /**
+     * Get a {@code Codec} using the given context, which includes, most importantly, the Class for which a {@code Codec} is required.
+     *
+     * <p>The default implementation delegates to {@link #get(Class, CodecRegistry)}, thus ignoring {@code typeArguments}.</p>
+     *
+     * @param clazz the Class for which to get a Codec
+     * @param typeArguments The type arguments for the {@code clazz}. The size of the list is either equal to the
+     * number of type parameters of the {@code clazz}, or is zero.
+     * For example, if {@code clazz} is {@link Collection}{@code .class}, then the size of {@code typeArguments} is one,
+     * since {@link Collection} has a single type parameter.
+     * The list may be {@linkplain List#isEmpty() empty} either because the {@code clazz} is not generic,
+     * or because another {@link CodecProvider} did not propagate {@code clazz}'s type arguments via the {@code registry},
+     * which may if that {@link CodecProvider} does not properly override {@link #get(Class, List, CodecRegistry)}.
+     * @param registry the registry to use for resolving dependent Codec instances
+     * @return the Codec instance, which may be null, if this source is unable to provide one for the requested Class
+     * @param <T> the type of the class for which a Codec is required
+     * @since 4.10
+     */
+    default <T> Codec<T> get(Class<T> clazz, List<Type> typeArguments, CodecRegistry registry) {
+        return get(clazz, registry);
+    }
```

Comment:
We don't have `Beta` in `bson`, but maybe we should. This will allow us to mark the new method as `@Beta(Beta.Reason.CLIENT)` and give ourselves more time to realize whether the signature of the method should be different. @jyemin What do you think?

Reply:
We were mistakenly happy previously more than once about things in this domain. Let's hope, this is the last time)